packages/babel-parser/src/plugins/estree.ts

```diff
@@ -121,6 +123,24 @@ export default (superClass: typeof Parser) =>
       return stmt;
     }
 
+    /**
+     * The TS-ESLint always define optional AST properties, here we provide the
+     * default value for such properties immediately after `finishNode` was invoked.
+     * This hook will be implemented by the typescript plugin.
+     *
+     * Note: This hook should be manually invoked when we change the `type` of a given AST
+     * node, to ensure that the optional properties are correctly filled.
+     * @param node The AST node finished by finishNode
+     */
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    fillOptionalPropertiesForTSESLint(node: NodeType) {}
+
+    castNodeAs<T extends N.Node>(node: N.Node, type: T["type"]): T {
```

Comment:
I don't know if this works, but maybe we can change the signature so that the type argument can be inferred?

```ts
castNodeTo<T extends N.Node>(type: T["type"], node: N.Node): T {
```
or
```ts
castNodeTo<const T extends N.Node["type"]>(node: N.Node, type: T): Extract<N.Node, { type: T }> {
```
or
```ts
castNodeTo<const T extends N.Node["type"]>(type: T, node: N.Node): Extract<N.Node, { type: T }> {
```

(or maybe without the `const` modifier)